dagger/runtime/cli/invoke_with_locations.py

```diff
@@ -83,8 +84,11 @@ def _validate_inputs(
     input_locations: Iterable[str],
 ):
     """Validate all the input locations supplied against the inputs the node is expecting."""
-    for input_name in input_names:
-        if input_name not in input_locations:
+    for input_name, input_value in input_names.items():
```

Comment:
You've refactored this to inject a `Mapping[str, Union[...]]` instead of an `Iterable[str]`, but you haven't changed the function's signature.

That said, I would keep the function as it was before because I think it's simpler and more testable, and instead of supplying the names of all the node's inputs, I would supply only the names of the required inputs.

I'd suggest creating a utility function named `filter_required_inputs` somewhere (ideally, next to wherever we place the `validate_parameters` function), that would return all the required inputs of a node.

Then, both this function and the `validate_parameters` function would use the `filter_required_inputs` directly.

This would also mean that we have a single source of truth for what "required" means.

Reply:
Makes sense! I've done it